carbonserver/carbonserver.go

```diff
@@ -826,6 +896,23 @@ func (listener *CarbonserverListener) expandGlobs(ctx context.Context, query str
 		fallbackToFS = true
 	}
 
+	if useGlob || fallbackToFS {
+		// no index or we were asked to hit the filesystem
+		for _, g := range globs {
+			nfiles, err := filepath.Glob(listener.whisperData + "/" + g)
+			if err == nil && nfiles != nil {
+				files = append(files, nfiles...)
+			} else {
+				// when cache-scan is enabled in the config, fallback to
+				// trigram-searching code when there is no whisper file
+				// to see if the metric exists only in cache
+				if listener.cacheGetRecentMetrics != nil{
+					useGlob = false
```

Comment:
note: trigram index isn't very efficient when it comes to searching full-named metric, i.e. not containing any pattern matching chars (*, [], {}, etc).

And would this causing duplicated search result?

Reply:
The intention here is to do a filepath glob first *and then* to fallback to trigram query when `cache-scan` is enabled. This way, for metrics with wsp files on disk, we only hit the filepath glob, and for cache-only metrics, we will also query the trigram index.

Reply:
Files exists on disk could also be indexed in trigram index, right? So if there the globs has miss match in filepath.Glob, then useGlob is set to false. And there is another `for _, g := range globs ` when querying trigram index bellow?

Reply:
Actually, if the there is malformed glob pattern in the globs, it could caused duplicated results because now you would range globs twice here and below.

I am not sure if this change is necessary? If you want to force go-carbon to always use trigram index in order to hit the still in-memory new metrics, maybe it's better to just force set useGlob to false?

Reply:
Oh! I see. Agreed. I will change the logic to force set useGlob to false when cache-scan is enabled.